lib/include/tick/base/report.h

```diff
@@ -0,0 +1,51 @@
+//
+//  report.h
+//  TICK
+//
+//  Created by Philip Deegan
+//  Copyright (c) 2015 bacry. All rights reserved.
+//
+//  Multipurpose Wrapper return type
+
+#ifndef LIB_INCLUDE_TICK_BASE_REPORT_H_
+#define LIB_INCLUDE_TICK_BASE_REPORT_H_
+
+#include <sstream>
+
+template <class A = bool, class B = std::string>
+class Report {};
+
+class BoolStrReport : public Report<bool, std::string> {
+  template <class T1>
+  friend std::ostream& operator<<(std::ostream&, const BoolStrReport&);
+
+ private:
+  bool success = false;
+  std::string str;
+
+ public:
+  BoolStrReport(bool b, const std::string& s) : success(b), str(s) {}
+  BoolStrReport(BoolStrReport& b, const std::string& s)
+      : success(b.success), str(s) {}
+  explicit BoolStrReport(const std::pair<bool, std::string>& p)
+      : success(p.first), str(p.second) {}
+  explicit BoolStrReport(const std::pair<BoolStrReport, std::string>& p)
+      : success(p.first.success), str(p.second) {}
+  BoolStrReport(const BoolStrReport& r)
+      : success(r.success), str(std::move(r.str)) {}
+  BoolStrReport(const BoolStrReport&& r)
+      : success(r.success), str(std::move(r.str)) {}
+  BoolStrReport& operator=(const BoolStrReport& r) = delete;
+  BoolStrReport& operator=(const BoolStrReport&& r) = delete;
+
+  std::string why() { return str; }
+
+  explicit operator bool() const { return success; }
+};
+
+template <typename T>
+inline std::ostream& operator<<(std::ostream& s, const BoolStrReport& r) {
+  return s << r.str;
```

Comment:
Can't we use r.why() and avoid friend function? 
I would still love to have `len(variable_name) > 3` :)